### PR TITLE
Support server-side-apply creation of new objects

### PR DESCRIPTION
**Pull request by justinsb**
This requires effectively a new code path in the update logic.

### CODE REVIEW

File: porch/Makefile [hunk 1]
```diff
@@ -163,6 +163,11 @@ build-images:
 	IMAGE_NAME="$(PORCH_FUNCTION_RUNNER_IMAGE)" WRAPPER_SERVER_IMAGE_NAME="$(PORCH_WRAPPER_SERVER_IMAGE)" make -C func/ build-image
 	IMAGE_NAME="$(TEST_GIT_SERVER_IMAGE)" make -C test/ build-image
 
+.PHONY: dev-server
```
Comment (martinmaly) [inline porch/Makefile:166]: nice!

File: porch/pkg/registry/porch/packagecommon.go [hunk 2]
```diff
@@ -141,30 +145,71 @@ func (r *packageCommon) updatePackageRevision(ctx context.Context, name string,
 
 	oldPackage, err := r.getPackage(ctx, name)
 	if err != nil {
-		return nil, false, err
+		if apierrors.IsNotFound(err) {
+			// For server-side apply, we create new objects where they don't already exist
```
Comment (martinmaly) [inline porch/pkg/registry/porch/packagecommon.go:149]: should we be checking `forceAllowCreate` here to see if creation on update is disabled?

(I honestly don't know whether the `false` value means "do not create on update" or  whether it means "do not force" which seems to read very permissive "just do whatever :)")

File: porch/pkg/registry/porch/packagecommon.go [hunk 3]
```diff
@@ -141,30 +145,71 @@ func (r *packageCommon) updatePackageRevision(ctx context.Context, name string,
...
+		} else {
+			return nil, false, err
+		}
+	}
+
+	isCreate := false
```
Comment (martinmaly) [inline porch/pkg/registry/porch/packagecommon.go:155]: [nit] I found this line surprising given 149 determined that we may proceed to creation.

I'd probably go for:

var isCreate bool
if ... {
  isCreate = true
} else {
}

or add a comment, or move the definition above line 146

File: porch/pkg/registry/porch/packagecommon.go [hunk 4]
```diff
@@ -141,30 +145,71 @@ func (r *packageCommon) updatePackageRevision(ctx context.Context, name string,
...
+				klog.Infof("update failed create validation: %v", err)
+				return nil, false, err
+			}
+		}
+
+		// TODO: ValidateCreate function ?
+		// fieldErrors := r.updateStrategy.ValidateCreate(ctx, newRuntimeObj, oldRuntimeObj)
+		// if len(fieldErrors) > 0 {
+		// 	return nil, false, apierrors.NewInvalid(api.SchemeGroupVersion.WithKind("PackageRevision").GroupKind(), name, fieldErrors)
+		// }
```
Comment (martinmaly) [inline porch/pkg/registry/porch/packagecommon.go:210]: or, like core k8s we should have CreateStrategy interface
[context](https://github.com/kubernetes/apiserver/blob/master/pkg/registry/rest/create.go#L42)
[k8s code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L564-L575)

**Review by martinmaly — APPROVED** [review]
lgtm with some nits. IMO cleanest approach would be to add
`SimpleRESTCreateStrategy` similar to https://github.com/GoogleContainerTools/kpt/blob/main/porch/pkg/registry/porch/rest.go#L26

File: porch/pkg/registry/porch/packagecommon.go (hunk 2 above)
Comment (justinsb) [reply porch/pkg/registry/porch/packagecommon.go]: Good idea - I added a check for it, and it seems to work as expected.

I agree it's not very clear what the intent of the flag is...

File: porch/pkg/registry/porch/packagecommon.go (hunk 3 above)
Comment (justinsb) [reply porch/pkg/registry/porch/packagecommon.go]: Good point; reorganized the code to clean this up

File: porch/pkg/registry/porch/packagecommon.go (hunk 4 above)
Comment (justinsb) [reply porch/pkg/registry/porch/packagecommon.go:211]: That's #3090 so I'm going to merge this one!

**Pull request merged**